---
title: "subtyper: ppmi subtyping with functional enrichment via SNPs"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{subtyper: ppmi subtyping with functional enrichment via SNPs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE,echo=FALSE,eval=TRUE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
set.seed( 000 )
fixmat <- function( x ) {
  xt = t( x )
  zz=apply( xt, FUN=var, MARGIN=2)
  x = t( xt[, zz != 0 ] )
  zz=apply( x, FUN=var, MARGIN=2)
  myrm = rowMeans( x, na.rm=T )
  for ( rr in which( zz == 0 ) )
    x[,rr]=myrm
  return( x )
}
library( gprofiler2 )
library( NMF )
library( subtyper )

```

# Functional enrichment


Single nucleotide polymorphisms (SNPs) are the most common type of genetic
variation among people and represent a difference, within or between genes,
in a sigle base pair.  SNPs are a "digital code" for genetic patterns.
Consequently, they are a very fine-grained, noisy measurement of genetic
variability.

Functional enrichment analysis uses meta-analysis to associate high-level
properties such as disease risk, high expression within given cell types or
other phenotypes with genes or SNPs.  This vignette uses an open dataset
[Nalls 2019](https://pubmed.ncbi.nlm.nih.gov/31701892/) and
gene ontology query tool (`ggprofiler`) to identify clusters of SNPs
associated with pleiotropic properties.  See [this review](https://www.frontiersin.org/articles/10.3389/fgene.2020.00424/full) for conceptual background on functional genomics.


First, we generate the association data.

```{r functionalenrichmentAll,echo=FALSE,eval=TRUE,fig.width=24,fig.height=8,message=FALSE}
myq = c( "rs10005233", "rs10119", "rs10784461", "rs1132899", "rs115729781", "rs11729289", "rs1372520", "rs157584", "rs184017", "rs2404832", "rs2638262", "rs28440052", "rs356165", "rs393584", "rs394221", "rs4072037", "rs5112", "rs584007", "rs6599388", "rs6859", "rs7134408", "rs71352239", "rs7305344", "rs74467833", "rs865961923" )
myq = c( "rs115729781", "rs394221"  ,  "rs157584" )
myq = c( "rs7305344",   "rs7134408" ,    "rs865961923" )
myq = c( "rs115729781", "rs394221" ,   "rs157584",    "rs7305344",   "rs7134408"  , "rs865961923")
myThresh = 0.01
thresh2 = 0.05
myCorr = 'g_SCS'
myCorr = 'fdr'
significant = FALSE
gostresall <- gost(query = myq,
  organism = "hsapiens", ordered_query = FALSE,
  multi_query = FALSE, significant = significant, exclude_iea = FALSE,
  measure_underrepresentation = FALSE, evcodes = FALSE,
  user_threshold = myThresh, correction_method = myCorr,
  domain_scope = "annotated", custom_bg = NULL,
  numeric_ns = "")#, sources = "HP" )

p <- gostplot(gostresall, capped = FALSE, interactive = FALSE)
print( p )
```

```{r functionalenrichmentBySNP,echo=FALSE,eval=TRUE,fig.width=24,fig.height=96,message=FALSE}
assocall = sort( unique( gostresall$result$term_name ) )
off = 0
assocK = list()
myviz = matrix( 0, nrow = length( myq ), ncol = length( assocall ) )
rownames(myviz)=myq
colnames(myviz)=assocall
for ( k in 1:length( myq ) ) {
  krange = (k-off):(k+off)
  krange = krange[ krange > 0 ]
  krange = krange[ krange <= length(myq) ]
  gostres <- gost(query = myq[krange],
    organism = "hsapiens", ordered_query = FALSE,
    multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
    measure_underrepresentation = FALSE, evcodes = FALSE,
    user_threshold = thresh2, correction_method = myCorr,
    domain_scope = "annotated", custom_bg = NULL,
    numeric_ns = ""  )

  assoc = gostres$result$term_name
  if ( ! is.null( assoc ) ) {
    assocp = gostres$result$p_value
    assocp = assocp[ assoc %in%  assocall ]
    assoc = assoc[ assoc %in%  assocall ]
    if ( length( assoc ) > 0  ) {
      ptx = log( assocp )
      myviz[ k, assoc ] =
        myviz[ k, assoc ] + ptx
      }
    }
  assocK[[k]] = assoc
  grepubs = grep("ubst",assoc)
  grepdopa = grep("opa",assoc)
  grepcer = grep("ereb",assoc)
  grepmorph = grep("ynucle",assoc)
  if ( length( grepdopa) > 0 | length( grepcer ) > 0 | length( grepubs ) > 0 | length( grepmorph ) > 0 ) {
#    print( paste(k,myq[k] ) )
#    print( assoc[ c(grepubs,grepdopa,grepcer,grepmorph) ] )
    }
  }
assocK = unique( sort( unlist( assocK ) ) )



myviz2 = matrix( 0, nrow = length( myq ), ncol = length( assocK ) )
rownames(myviz2)=myq
colnames(myviz2)=assocK
for ( k in 1:length( myq ) ) {
  krange = (k-off):(k+off)
  krange = krange[ krange > 0 ]
  krange = krange[ krange <= length(myq) ]
  gostres <- gost(query = myq[krange],
    organism = "hsapiens", ordered_query = FALSE,
    multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
    measure_underrepresentation = FALSE, evcodes = FALSE,
    user_threshold = thresh2, correction_method = myCorr,
    domain_scope = "annotated", custom_bg = NULL,
    numeric_ns = "", sources = NULL )
  assoc = gostres$result$term_name
  if ( ! is.null( assoc ) ) {
    assocp = gostres$result$p_value
    assocp = assocp[ assoc %in%  assocK  ]
    assoc=assoc[ assoc %in%  assocK  ]
    if ( length( assoc ) > 0  ) {
      ptx = log( assocp )
      myviz2[ k, assoc ] = myviz2[ k, assoc ] + ptx
      }
    }
  }
myviz2 = myviz2[ , abs(colSums(myviz2)) > 0 ]
# pheatmap::pheatmap( abs(myviz), cluster_cols=F, cluster_rows=F, show_rownames=F, show_colnames=F )
# pdf("~/Downloads/temp/temp.pdf",width=32,height=96)
pheatmap::pheatmap( abs(t(myviz2)), cluster_cols=T, cluster_rows=T, show_rownames=T, show_colnames=T )
# dev.off()

gsnpenseres = gsnpense( query = myq, filter_na = TRUE )

```

Next, we explicitly perform clustering.  First, cluster based on SNPs.

```{r stfeasnpclusters,echo=FALSE,eval=TRUE}
bcdata = -1.0 * t( myviz2 )
bcdata[ is.na( bcdata ) ] = 0.0
myclusts = trainSubtypeClusterMulti( bcdata,
  colnames(bcdata), method = "kmeans",  desiredk = 3 )
bcdatac = predictSubtypeClusterMulti( bcdata, colnames(bcdata), myclusts,
  clustername='FUNCcluster'  )
```


Second, cluster based on function.

```{r stfeagoclusters,echo=FALSE,eval=TRUE}
bcdata2 = 1.0 * ( myviz2 )
bcdata2[ is.na( bcdata2 ) ] = 0.0
maxk = min( c( nrow(bcdata2) - 2, round(nrow( bcdata2 )/2), 10 ) )
myclusts2 = trainSubtypeClusterMulti( bcdata2,
  colnames(bcdata2), method = "kmeans",  maxk = maxk )
bcdata2c = predictSubtypeClusterMulti( bcdata2, colnames(bcdata2), myclusts2,
  clustername='SNPcluster' )
```


## bi-clustering via joint clustering and via NMF

* tabulate the pairs of clusters that exist

* provide evidence that joint clusters are valid (TODO)

```{r simple,echo=FALSE,eval=TRUE,fig.width=12,fig.height=8}
mysnps = rownames( myviz2 )
myfunc = colnames( myviz2 )
jointClusterMatrix = matrix( "", nrow = nrow(myviz2), ncol = ncol(myviz2) )
colnames( jointClusterMatrix ) = colnames( myviz2 )
rownames( jointClusterMatrix ) =
for ( i in 1:nrow(myviz2) ) {
  mysnp = mysnps[i]
  snpcluster = bcdata2c$SNPcluster[ rownames( bcdata2c ) == mysnp ]
  for ( j in 1:ncol(myviz2) ) {
    myfnc = myfunc[j]
    fnccluster = bcdatac$FUNCcluster[ rownames( bcdatac ) == myfnc ]
    myclustid = paste0( snpcluster, '-',  fnccluster )
    jointClusterMatrix[i,j]=myclustid
  }
}
######## show a resulting matrix
mytbl = table( jointClusterMatrix )
knitr::kable( mytbl, caption='Joint cluster counts.')

cnames = names( mytbl )
binmat = myviz2 * 0
ct = 1
for ( nm in cnames ) {
  binmat[ jointClusterMatrix == nm ] = ct
  ct = ct + 1
}
pheatmap::pheatmap( myviz2, cluster_cols=T, cluster_rows=T, show_rownames=F, show_colnames=F)
pheatmap::pheatmap( binmat, cluster_cols=T, cluster_rows=T, show_rownames=F, show_colnames=F)

# pheatmap::pheatmap( jointClusterMatrix , cluster_cols=F, cluster_rows=F, show_rownames=F, show_colnames=F)
```

```{r eanat,echo=FALSE,eval=FALSE}
library( ANTsR )
nv = 10
eanatmat = antsrimpute( scale( abs(data.matrix( fixmat(myviz2) )), T,T) )
eanatmat = scale( abs(data.matrix( fixmat(myviz2) )), T,T)
inmats = list( eanatmat, eanatmat )
eanat = simlr( inmats, initialUMatrix = nv, sparsenessQuantiles=rep(0.9,nv),
  mixAlg = 'ica', energyType = 'regression' )
# eanat = sparseDecom( eanatmat, sparseness = 0.1, nvecs = nv,
#  its = 25, z = 0.2, mycoption = 0 )
qq = 0.5
for ( j in 1:nv ) {
  uvec = abs( eanat$u[[1]][,j] )
  vvec = abs( eanat$v[[1]][,j] )
  uvec = uvec / sum( uvec )
  vvec = vvec / sum( vvec )
  snps1=rownames( eanatmat )[ uvec > quantile(uvec,qq)  & uvec > 1e-9 ]
  func1=colnames( eanatmat )[ vvec > quantile(vvec,qq)  & vvec > 1e-9 ]
  print( j )
  print( snps1 )
  print( func1 )
  subviz = eanatmat[ rownames( eanatmat ) %in% snps1 ,  colnames( eanatmat ) %in% func1  ]
  pheatmap::pheatmap( abs(subviz), cluster_cols=F, cluster_rows=T, show_rownames=T, show_colnames=F )
  Sys.sleep( 3 )
}


```


```{r nmf,echo=FALSE,eval=TRUE,fig.width=12,fig.height=8}
mybic = biclusterMatrixFactorization( myviz2 * ( -1.0 ), colnames(myviz2), k = 5 )
colnames(mybic$jointClusters)=substr( colnames(mybic$jointClusters), 0, 12 )
pheatmap::pheatmap( mybic$jointClusters, cluster_cols=F, cluster_rows=T, show_rownames=T, show_colnames=T )
```


```{r,echo=FALSE,eval=FALSE}
library( biclust )
eanatmat = antsrimpute( scale( abs(data.matrix( fixmat(myviz2) )), T,T) )
mybc = biclust( x = data.matrix( eanatmat ), method = BCPlaid() )
heatmapBC( x = eanatmat, bicResult = mybc )
#
# biclustbarchart(bcdata,mybc, col="#A3E0D8")
# ord<-bicorder( data.frame(bcdata), cols=TRUE, rev=TRUE)
# biclustbarchart( bcdata, mybc )

```



```{r,echo=FALSE,eval=FALSE}
library( sparseBC )
myspbc = sparseBC( bcdata, 3, 3, 0.1 )
# Display some information on the object sparseBC
summary( myspbc )
# Image plots to illustrate the estimated mean matrix
image(t(myspbc$mus),main="sparseBC")

```



## make a heat map of the brain areas associated with SNPs

```{r brainmap,echo=FALSE,eval=FALSE}

citfn=path.expand("~/.antspyt1w/CIT168_Reinf_Learn_v1_label_descriptions_pad.csv")
dktfn = path.expand("~/.antspyt1w/dkt.csv")
dkt = read.csv( dktfn )
dkt$Description = gsub(" ","_", dkt$Description)
cit = read.csv( citfn )
mediaDF = read.csv(path.expand("~/Downloads/temp/debug_NBM/nihcardassembly/ppmi/csvs/mediaDF.csv"))

sel1a = mediaDF$mediation &
  mediaDF$outcome == "UPDRS.Part1.Patient.Completed.Total.UPDRS1.Num"
sel1b = mediaDF$mediation &
  mediaDF$outcome == "UPDRS.Part1.Rater.Completed.Total.UPDRS1.Num"
sel2 = mediaDF$mediation & mediaDF$outcome == "PIGD.Subscore.UPDRS3"
a1a=unique( mediaDF[ sel1a, "anatname" ] )
a1b=unique( mediaDF[ sel1b, "anatname" ] )
a2=unique( mediaDF[ sel2, "anatname" ] )
a3 = intersect( a1b, a2 )
s1a=unique( mediaDF[ sel1a, "snp" ] )
s1b=unique( mediaDF[ sel1b, "snp" ] )
s2=unique( mediaDF[ sel2, "snp" ] )
s3 = intersect( s1b, s2 )
# table( a1 %in% a2 )
# table( s1 %in% s2 )

library(ANTsR)
ifn = path.expand("/Users/stnava/data/UKBB/example_data/processed/UKBB/1619189/2/T1wHierarchical/20252/UKBB_1619189_2_T1wHierarchical_20252_brain_n4_dnz.nii.gz")
img=antsImageRead( ifn )
sfn = path.expand("/Users/stnava/data/UKBB/example_data/processed/UKBB/1619189/2/T1wHierarchical/20252/UKBB_1619189_2_T1wHierarchical_20252_dkt_cortex.nii.gz")
seg=antsImageRead( sfn )

sfn = path.expand("/Users/stnava/data/UKBB/example_data/processed/UKBB/1619189/2/T1wHierarchical/20252/UKBB_1619189_2_T1wHierarchical_20252_tissue_segmentation.nii.gz")
tseg=antsImageRead( sfn )

 wm = thresholdImage( tseg, 3, 4) %>% iMath("FillHoles")
 wm = wm  %>% thresholdImage(1, 2)
 wms = smoothImage( wm, 1.0 )
 wmk = weingartenImageCurvature( wms, 1.5, 'mean' )
 # will display to screen
 myantspath = '/opt/ANTs/bin/'
 rp1 = matrix( c(90,180,90), ncol = 3 )
 rp2 = matrix( c(270, 0, 90), ncol = 3 )
 rp  = rbind( rp1, rp2 )
 wm2 = ( antsImageClone( wm ) )
 wm2[1:(304/2),1:354,1:296]=0

regions2seg <- function( searcher ) {
  a1asearch = gsub("vol__","",searcher)
  a1asearch = gsub("thk__","",a1asearch)
  a1asearch = gsub("dktcortex_SR","",a1asearch)
  a1asearch = unique( gsub("dktregions_SR","",a1asearch))
  dktL = dkt[ grep("right",dkt$Description),]
  witcher = grep( paste0(a1asearch,collapse='|'), dktL$Description)
  seg1labs = dktL$Label[ witcher ]
  seg1 = maskImage( seg, seg, level=seg1labs, TRUE)
  return( seg1 )
}

antsImageWrite( img, '~/Downloads/tempi.nii.gz')
seg1 = regions2seg( a1a )
antsImageWrite( seg1, '~/Downloads/temps1a.nii.gz')
seg1b = regions2seg( a1b )
antsImageWrite( seg1b, '~/Downloads/temps1b.nii.gz')
seg2 = regions2seg( a2 )
antsImageWrite( seg2, '~/Downloads/temps2.nii.gz')


snps2map = unique( mediaDF[ mediaDF$mediation , "snp"] )
for ( s in snps2map ) {
  a1a=unique( mediaDF[ sel1a, "anatname" ] )
  subber = mediaDF[ mediaDF$mediation & mediaDF$snp == s, ]
  for ( outc in unique( subber$outcome ) ) {
    myanat = unique( mediaDF[ mediaDF$mediation &
      mediaDF$snp == s & mediaDF$outcome == outc, 'anatname' ] )
    wfn = paste0('/Users/stnava/Downloads/temp/debug_NBM/nihcardassembly/ppmi/figs/ppmi_', outc, "_", s, "_surf" )
    print( wfn )
    print( myanat )
    blob = regions2seg( myanat )
    sblob = smoothImage(blob,3) %>%  iMath("Normalize")
    z = list( thresholdImage(sblob, 0.5,1) )
    y = list( sblob )
    antsrSurf( x=wm2,
      z = z,
      y = y,
      colormap=c("red"), alpha=c(1),
      rotationParams = rp,
      overlayLimits = list( c(0.15,1.00001)),
      # smoothingSigma=0.5,
      inflationFactor=256,
      filename = wfn,
      antspath=myantspath )

#    antsImageWrite( seg2, wfn )
  }
}

```




```{r antssurfex,echo=FALSE,eval=FALSE}

# antsrSurf( x=wm, y = list( wmk %>% smoothImage(1)), z=list( wm %>% iMath("MD",1)),
#  rotationParams = c(270, 0, 90), overlayLimits=c(-0.4,0.4), antspath=myantspath )



# antsrSurf( x=wm, y=y, z=z, smoothingSigma=0.5, alpha=c( 1, 1, 1),
#     colormap=c("red","blue"),
#     inflationFactor=255, overlayLimits=list( c(0.5,1.0001), c(0.5,1.0001) ),
#     verbose = TRUE, rotationParams = rp[1,], antspath=myantspath )
```
